### PR TITLE
librustc_trans: Handle DST structs in trans::_match.

### LINKED ISSUE #23261: ICE when matching on DST structs.
The following code causes 

    Assertion failed: (getOperand(0)->getType() == cast<PointerType>(getOperand(1)->getType())->getElementType() && "Ptr must be a pointer to Val type!"), function AssertOK, file /Users/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-mac/build/src/llvm/lib/IR/Instructions.cpp, line 1083.

```rust
use std::collections::HashMap;
use std::path::{Path, PathBuf};

#[derive(Debug)]
pub struct FileMeta;

struct S {
    h2 : HashMap<PathBuf, FileMeta>,
}

impl S {
    fn new() -> S {
        S {
            h2: HashMap::new(),
        }
    }

    fn f(&self, path: &Path) -> FileMeta {
        let m = FileMeta;
        match self.h2.get(path) {
            Some(meta) => m,
            None => m
        }
    }
}
```

Using `rustc 1.0.0-nightly (12b846ab8 2015-03-09) (built 2015-03-09)`.

### PR DESCRIPTION
**Pull request by luqmana**
Fixes #23261.

cc @blaenk

### CODE REVIEW

**Comment by huonw** [comment]
Does this handle trait objects?

**Comment by luqmana** [comment]
@huonw Yes, it should. Let me update the test case.

File: src/librustc_trans/trans/_match.rs [hunk 1]
```diff
@@ -1076,9 +1077,35 @@ fn compile_submatch_continue<'a, 'p, 'blk, 'tcx>(mut bcx: Block<'blk, 'tcx>,
     let adt_vals = if any_irrefutable_adt_pat(bcx.tcx(), m, col) {
         let repr = adt::represent_type(bcx.ccx(), left_ty);
         let arg_count = adt::num_args(&*repr, 0);
-        let field_vals: Vec<ValueRef> = (0..arg_count).map(|ix|
-            adt::trans_field_ptr(bcx, &*repr, val, 0, ix)
+        let (arg_count, struct_val) = if type_is_sized(bcx.tcx(), left_ty) {
+            (arg_count, val)
+        } else {
+            (arg_count - 1, Load(bcx, expr::get_dataptr(bcx, val)))
```
Comment (huonw) [inline src/librustc_trans/trans/_match.rs]: Could you add a comment about the - 1?

**Comment by huonw** [comment]
r=me with an explanatory comment

**Comment by blaenk** [comment]
Great job @luqmana :+1:

**Pull request merged**